File: 1535-find-the-winner-of-an-array-game/1535-find-the-winner-of-an-array-game.py
class Solution:
    def getWinner(self, arr: List[int], k: int) -> int:
        max_element = max(arr)
        curr, winstreak = arr[0], 0
        
        for i in range(1, len(arr)):
            opponent = arr[i]
            if curr > opponent:
                winstreak += 1
            else:
                curr = opponent
                winstreak = 1
            
            if winstreak == k or curr == max_element:
                return curr
        return -1

File: 1535-find-the-winner-of-an-array-game/1535-find-the-winner-of-an-array-game.cpp
class Solution {
public:
    int getWinner(vector<int>& arr, int k) {
        int mx = *max_element(arr.begin(), arr.end()), curr = arr[0], winstreak = 0, opponent;
        queue<int> queue({arr.begin() + 1, arr.end()});
        
        while (!queue.empty()){
            opponent = queue.front();
            queue.pop();
            
            if (curr > opponent){
                queue.push(opponent);
                ++winstreak;
            }
            else{
                queue.push(curr);
                curr = opponent;
                winstreak = 1;
            }
            
            if (curr == mx || winstreak == k)
                return curr;
        }
        
        return -1;
    }
};